src/simulator.cu

```diff
@@ -11,208 +11,259 @@
+    for (int i = 0; i < params.numParticles; i++) {
+      float t = dis(gen) * 2 * PI;
+      float s = dis(gen) * 100;
+      pos.x[i] = cos(t) * s;
+      pos.y[i] = sin(t) * s;
+    }
+
+    // Z component is independent (uniform range 0-4)
+    std::generate(begin(pos.z), end(pos.z),
+        [&gen, &dis]() { return 4.0 * dis(gen); });
+  }
+
+  void DiskGalaxySimulator::initialParticleVel() {
+    for (int i = 0; i < params.numParticles; i++) {
+      vec3 vel = cross({pos.x[i], pos.y[i], pos.z[i]}, {0.0, 0.0, 1.0});
+      coords_t orbital_vel = std::sqrt(2.0 * length(vel));
+      vel = normalize(vel) * orbital_vel;
+      this->vel.x[i] = vel.x;
+      this->vel.y[i] = vel.y;
+      this->vel.z[i] = vel.z;
+    }
+  }
+
+  const ParticleData& DiskGalaxySimulator::getParticlePos() { return pos; };
+
+  const ParticleData& DiskGalaxySimulator::getParticleVel() { return vel; };
+
+  // Linear Algebra functions (not yet exposed in header)
+  HOSTDEV vec3 cross(const vec3 v0, const vec3 v1) {
+    return vec3(v0.y * v1.z - v0.z * v1.y, v0.z * v1.x - v0.x * v1.z,
+        v0.x * v1.y - v0.y * v1.x);
+  };
+
+  HOSTDEV coords_t length(const vec3 v) {
+    return std::sqrt(std::pow(v.x, 2) + std::pow(v.y, 2) + std::pow(v.z, 2));
+  }
+
+  HOSTDEV vec3 normalize(const vec3 v) {
+    vec3 result = v;
+    coords_t len = length(v);
+    result.x /= len;
+    result.y /= len;
+    result.z /= len;
+    return result;
+  }
+
+  /* O(n^2) implementation (no distance threshold), with no shared
+     memory etc.
    */
-   __global__ void particle_interaction(ParticleData_d pPos,
-                                        ParticleData_d pNextPos,
-                                        ParticleData_d pVel, SimParam params) {
-      int id = threadIdx.x + (blockIdx.x * blockDim.x);
-      if (id >= params.numParticles) return;
-
-      vec3 force(0.0f, 0.0f, 0.0f);
-      vec3 pos(pPos.x[id], pPos.y[id], pPos.z[id]);
-
-      #pragma unroll 4
-      for (int i = 0; i < params.numParticles; i++) {
-         vec3 other_pos{pPos.x[i], pPos.y[i], pPos.z[i]};
-         vec3 r = other_pos - pos;
-         // Fast computation of 1/(|r|^3)
-         coords_t dist_sqr = dot(r, r) + params.distEps;
-         coords_t inv_dist_cube = rsqrt(dist_sqr * dist_sqr * dist_sqr);
-
-         // assume uniform unit mass
-         force += r * inv_dist_cube * (i != id);
-      }
-
-      // Update velocity
-      vec3 curr_vel(pVel.x[id], pVel.y[id], pVel.z[id]);
-      curr_vel *= params.damping;
-      curr_vel += force * params.dt * params.G;
-
-      pVel.x[id] = curr_vel.x;
-      pVel.y[id] = curr_vel.y;
-      pVel.z[id] = curr_vel.z;
-
-      // Update position (integration)
-      vec3 curr_pos(pPos.x[id], pPos.y[id], pPos.z[id]);
-
-      curr_pos += curr_vel * params.dt;
-      pNextPos.x[id] = curr_pos.x;
-      pNextPos.y[id] = curr_pos.y;
-      pNextPos.z[id] = curr_pos.z;
-   }
-
+  __global__ void particle_interaction_b(ParticleData_d pPos,
```

Comment:
We could reduce the code duplication a bit here with something like:

```c++
enum class CalculationMethod {
  BRANCH,
  PREDICATED
};

template <CalculationType ct>
__global__ void particle_interaction_b(ParticleData_d pPos, // &c
...
  if constexpr(ct == CalculationType::BRANCH) {
    if (id == i) continue;
    force += r * inv_dist_cube;
  } else if constexpr (ct == CalculationType::PREDICATED)
    // ... and so on
```

However, this isn't needed, as it will still involve the same number of kernels being generated, but will have a bit less code. This is roughly what I sketched when we talked about this last week.